File: langs/csharp/src/VMx/Hierarchical/HierarchicalVMBuilder.cs
#pragma warning disable CA1715 // Spec uses 'TModel' / 'TVM' per ADR-0006 and ADR-0028
using VMx.Builders;
using VMx.Services;

namespace VMx.Hierarchical;

/// <summary>
/// Construction context passed to <see cref="HierarchicalVMBuilder{TModel, TVM}.VmFactory"/>.
/// Carries the validated constructor arguments produced by the builder so consumers
/// can wire their concrete <typeparamref name="TVM"/> subclass without having to
/// inspect the builder directly.
/// </summary>
/// <typeparam name="TModel">Domain model carried by the node.</typeparam>
/// <typeparam name="TVM">Concrete subclass type (recursive constraint per ADR-0028 §3.2).</typeparam>
/// <param name="Model">Domain model for the node being built.</param>
/// <param name="ChildrenFactory">Factory invoked with the new node to produce its children.</param>
/// <param name="Hub">Message hub for pub/sub.</param>
/// <param name="Dispatcher">Dispatcher for async/background work.</param>
/// <param name="Name">Optional VM name (defaults to <c>typeof(TVM).Name</c> when null).</param>
/// <param name="Hint">Optional display hint string (defaults to empty).</param>
/// <param name="EagerChildren">When true, the full subtree is materialized at construct time.</param>
public sealed record HierarchicalVMConstructionContext<TModel, TVM>(
    TModel Model,
    Func<TVM, IEnumerable<TVM>> ChildrenFactory,
    IMessageHub Hub,
    IDispatcher Dispatcher,
    string? Name,
    string Hint,
    bool EagerChildren)
    where TVM : HierarchicalVM<TModel, TVM>;

/// <summary>
/// Immutable fluent builder for <see cref="HierarchicalVM{TModel, TVM}"/>.
/// Each setter returns a new builder instance (BLD-001).
/// See spec/10-builders.md §3 and ADR-0035 §2 H1 / H2.
///
/// Because <see cref="HierarchicalVM{TModel, TVM}"/> is <c>abstract</c> with a
/// <c>protected</c> constructor, consumers MUST supply a <see cref="VmFactory"/>
/// callback that knows how to instantiate their concrete subclass. The factory
/// receives a <see cref="HierarchicalVMConstructionContext{TModel, TVM}"/>
/// carrying the validated constructor arguments, e.g.
/// <code>
/// .VmFactory(ctx =&gt; new MyTreeNode(
///     ctx.Model, ctx.ChildrenFactory, ctx.Hub, ctx.Dispatcher,
///     ctx.Name, ctx.Hint, ctx.EagerChildren))
/// </code>
///
/// Note: there is no <c>WithDefaultServices()</c> Wither on the C# builder. Per
/// ADR-0035 §2 H2 that Wither is Python + TypeScript only — C# consumers must
/// call <see cref="Services"/> explicitly.
/// </summary>
/// <typeparam name="TModel">Domain model carried by the node.</typeparam>
/// <typeparam name="TVM">Concrete subclass type (recursive constraint per ADR-0028 §3.2).</typeparam>
public sealed class HierarchicalVMBuilder<TModel, TVM>
    where TVM : HierarchicalVM<TModel, TVM>
{
    // ── Required ──────────────────────────────────────────────────────────────
    // `TModel` may be a value type, so we cannot use `null` to detect "unset".
    // Track presence with a dedicated flag (mirrors Python's `_model_set` sentinel).
    private readonly TModel _model;
    private readonly bool _modelSet;
    private readonly Func<TVM, IEnumerable<TVM>>? _childrenFactory;
    private readonly IMessageHub? _hub;
    private readonly IDispatcher? _dispatcher;
    private readonly Func<HierarchicalVMConstructionContext<TModel, TVM>, TVM>? _vmFactory;

    // ── Optional ──────────────────────────────────────────────────────────────
    private readonly string? _name;
    private readonly string _hint;
    private readonly bool _eagerChildren;

    /// <summary>Empty starting builder.</summary>
    public static readonly HierarchicalVMBuilder<TModel, TVM> Empty = new();

    private HierarchicalVMBuilder()
    {
        _model = default!;
        _modelSet = false;
        _hint = "";
    }

    private HierarchicalVMBuilder(
        TModel model,
        bool modelSet,
        Func<TVM, IEnumerable<TVM>>? childrenFactory,
        IMessageHub? hub,
        IDispatcher? dispatcher,
        Func<HierarchicalVMConstructionContext<TModel, TVM>, TVM>? vmFactory,
        string? name,
        string hint,
        bool eagerChildren)
    {
        _model = model;
        _modelSet = modelSet;
        _childrenFactory = childrenFactory;
        _hub = hub;
        _dispatcher = dispatcher;
        _vmFactory = vmFactory;
        _name = name;
        _hint = hint;
        _eagerChildren = eagerChildren;
    }

    /// <summary>Sets the required <c>Model</c> for this node.</summary>
    public HierarchicalVMBuilder<TModel, TVM> Model(TModel model)
        => With(model: model, modelSet: true);

    /// <summary>Sets the required children factory (parent -> child sequence).</summary>
    public HierarchicalVMBuilder<TModel, TVM> ChildrenFactory(Func<TVM, IEnumerable<TVM>> factory)
        => With(childrenFactory: factory);

    /// <summary>Sets the required services (hub + dispatcher).</summary>
    public HierarchicalVMBuilder<TModel, TVM> Services(IMessageHub hub, IDispatcher dispatcher)
        => With(hub: hub, dispatcher: dispatcher);

    /// <summary>
    /// Sets the required concrete-VM factory. Because
    /// <see cref="HierarchicalVM{TModel, TVM}"/> is abstract, the builder needs
    /// to be told which concrete subclass to instantiate. The factory receives a
    /// <see cref="HierarchicalVMConstructionContext{TModel, TVM}"/> and returns
    /// the concrete <typeparamref name="TVM"/>.
    /// </summary>
    public HierarchicalVMBuilder<TModel, TVM> VmFactory(
        Func<HierarchicalVMConstructionContext<TModel, TVM>, TVM> factory)
        => With(vmFactory: factory);

    /// <summary>Sets the optional <c>Name</c> (default: <c>typeof(TVM).Name</c>).</summary>
    public HierarchicalVMBuilder<TModel, TVM> Name(string name) => With(name: name);

    /// <summary>Sets the optional <c>Hint</c> (default: empty string).</summary>
    public HierarchicalVMBuilder<TModel, TVM> Hint(string hint) => With(hint: hint);

    /// <summary>
    /// When <see langword="true"/>, materializes the entire subtree at construct
    /// time (depth-first). Default is <see langword="false"/> (lazy).
    /// </summary>
    public HierarchicalVMBuilder<TModel, TVM> EagerChildren(bool eagerChildren)
        => With(eagerChildren: eagerChildren);

    /// <summary>
    /// Validates required fields and builds a concrete <typeparamref name="TVM"/>
    /// via the supplied <see cref="VmFactory"/>.
    /// </summary>
    public TVM Build()
    {
        if (!_modelSet) throw new BuilderValidationException("Model");
        BuilderValidationException.Require(_childrenFactory, "ChildrenFactory");
        BuilderValidationException.Require(_hub, "Hub");
        BuilderValidationException.Require(_dispatcher, "Dispatcher");
        BuilderValidationException.Require(_vmFactory, "VmFactory");

        var ctx = new HierarchicalVMConstructionContext<TModel, TVM>(
            Model: _model,
            ChildrenFactory: _childrenFactory,
            Hub: _hub,
            Dispatcher: _dispatcher,
            Name: _name,
            Hint: _hint,
            EagerChildren: _eagerChildren);

        return _vmFactory(ctx);
    }

    private HierarchicalVMBuilder<TModel, TVM> With(
        TModel? model = default,
        bool? modelSet = null,
        Func<TVM, IEnumerable<TVM>>? childrenFactory = null,
        IMessageHub? hub = null,
        IDispatcher? dispatcher = null,
        Func<HierarchicalVMConstructionContext<TModel, TVM>, TVM>? vmFactory = null,
        string? name = null,
        string? hint = null,
        bool? eagerChildren = null)
        => new(
            modelSet == true ? model! : _model,
            modelSet ?? _modelSet,
            childrenFactory ?? _childrenFactory,
            hub ?? _hub,
            dispatcher ?? _dispatcher,
            vmFactory ?? _vmFactory,
            name ?? _name,
            hint ?? _hint,
            eagerChildren ?? _eagerChildren);
}
#pragma warning restore CA1715

File: langs/swift/Sources/VMx/Builders/HierarchicalVMBuilder.swift
//
// HierarchicalVMBuilder<TModel, TVM> — immutable fluent builder for
// HierarchicalVM<TModel, TVM>.
//
// See spec/10-builders.md §3 and spec/18-hierarchical-vm.md (HIER-015/016/017).
//
// vmFactory decision (ADR-0061 §2.2):
// TypeScript requires a vmFactory because JS erases generic-type identity at
// runtime. Swift has reified generics but binds `TVM: AnyObject` (not
// `TVM: HierarchicalVM<TModel, TVM>`) because the self-referential constraint
// is rejected by the Swift compiler (see HierarchicalVM.swift file header).
// `AnyObject` exposes no `init(...)` surface, so this builder also cannot
// construct TVM directly and requires the same vmFactory pattern as TypeScript.
// The divergence is documented in ADR-0061 §2.2.
//
import Foundation

/// Immutable fluent builder for `HierarchicalVM<TModel, TVM>`.
///
/// Required fields, validated in this order at `build()`:
///   1. `model`
///   2. `childrenFactory`
///   3. `services` (hub + dispatcher)
///   4. `vmFactory`
///
/// Optional fields: `name` (default: `""` → resolved by `HierarchicalVM.init`
/// to `String(describing: TVM.self)`), `hint` (default: `""`),
/// `eagerChildren` (default: `false`).
///
/// Each setter returns a **new** builder instance; the original is unchanged
/// (BLD-001 / Swift struct copy semantics).
public struct HierarchicalVMBuilder<TModel, TVM: AnyObject> {

    // ── Stored state ─────────────────────────────────────────────────────

    private var _model: ModelBox?
    private var _childrenFactory: ((TVM) -> [TVM])?
    private var _hub: MessageHubProtocol?
    private var _dispatcher: Dispatcher?
    private var _name: String = ""
    private var _hint: String = ""
    private var _eagerChildren: Bool = false
    private var _vmFactory: ((TModel, @escaping (TVM) -> [TVM], MessageHubProtocol, Dispatcher, String, String, Bool) -> TVM)?

    /// One-shot box to distinguish "model was set (possibly to a TModel zero
    /// value)" from "model was never set" without `Hashable` constraints or
    /// sentinel values.
    private struct ModelBox { let value: TModel }

    // ── Init ─────────────────────────────────────────────────────────────

    public init() {}

    // ── Setters (copy-mutate-return) ─────────────────────────────────────

    /// Set the required domain model carried by each produced node.
    public func model(_ value: TModel) -> HierarchicalVMBuilder<TModel, TVM> {
        var c = self; c._model = ModelBox(value: value); return c
    }

    /// Set the required factory that produces the node's children given the
    /// (CRTP) parent node.
    public func childrenFactory(
        _ factory: @escaping (TVM) -> [TVM]
    ) -> HierarchicalVMBuilder<TModel, TVM> {
        var c = self; c._childrenFactory = factory; return c
    }

    /// Set the required message hub + dispatcher pair.
    public func services(
        hub: MessageHubProtocol, dispatcher: Dispatcher
    ) -> HierarchicalVMBuilder<TModel, TVM> {
        var c = self; c._hub = hub; c._dispatcher = dispatcher; return c
    }

    /// Set the optional node name (default: `""` → resolved by
    /// `HierarchicalVM.init` to `String(describing: TVM.self)`).
    public func name(_ value: String) -> HierarchicalVMBuilder<TModel, TVM> {
        var c = self; c._name = value; return c
    }

    /// Set the optional hint string (default: `""`).
    public func hint(_ value: String) -> HierarchicalVMBuilder<TModel, TVM> {
        var c = self; c._hint = value; return c
    }

    /// Set whether to eagerly materialize the child subtree at construct()
    /// time (default: `false`).
    public func eagerChildren(_ value: Bool) -> HierarchicalVMBuilder<TModel, TVM> {
        var c = self; c._eagerChildren = value; return c
    }

    /// Set the required factory that instantiates the concrete `TVM` subclass.
    ///
    /// Swift cannot call `TVM.init(...)` directly because `TVM` is bound to
    /// `AnyObject` (the self-referential constraint
    /// `TVM: HierarchicalVM<TModel, TVM>` is rejected by the Swift compiler;
    /// see `HierarchicalVM.swift`). Consumers supply a factory:
    ///
    /// ```swift
    /// .vmFactory { model, cf, hub, disp, name, hint, eager in
    ///     MyNode(model: model, childrenFactory: cf, hub: hub,
    ///            dispatcher: disp, name: name, hint: hint,
    ///            eagerChildren: eager)
    /// }
    /// ```
    ///
    /// The `@escaping` attribute on the `cf` parameter within the setter type
    /// allows passing `cf` to `HierarchicalVM.init(childrenFactory:)` which
    /// requires an `@escaping` closure.
    public func vmFactory(
        _ factory: @escaping (TModel, @escaping (TVM) -> [TVM], MessageHubProtocol, Dispatcher, String, String, Bool) -> TVM
    ) -> HierarchicalVMBuilder<TModel, TVM> {
        var c = self; c._vmFactory = factory; return c
    }

    /// Wire a fresh `MessageHub` + `ImmediateDispatcher.INSTANCE` as the
    /// default services pair. Mirrors `withDefaultServices()` in TypeScript /
    /// Python (ADR-0035 §2 H2).
    public func withDefaultServices() -> HierarchicalVMBuilder<TModel, TVM> {
        services(hub: MessageHub(), dispatcher: ImmediateDispatcher.INSTANCE)
    }

    // ── build() ──────────────────────────────────────────────────────────

    /// Validate all required fields and construct a `TVM` node.
    ///
    /// Validation order: `model` → `childrenFactory` → `services` →
    /// `vmFactory`. The first missing field causes a `BuilderValidationError`.
    ///
    /// - Throws: `BuilderValidationError(missingField:)` naming the first
    ///   missing required field.
    /// - Returns: A freshly constructed `TVM` node with the builder's settled
    ///   configuration. Calling `build()` again produces an independent
    ///   instance sharing equivalent configuration (HIER-016).
    public func build() throws -> TVM {
        guard let box = _model else {
            throw BuilderValidationError(missingField: "model")
        }
        guard let childrenFactory = _childrenFactory else {
            throw BuilderValidationError(missingField: "childrenFactory")
        }
        guard let hub = _hub, let dispatcher = _dispatcher else {
            throw BuilderValidationError(missingField: "services")
        }
        guard let vmFactory = _vmFactory else {
            throw BuilderValidationError(missingField: "vmFactory")
        }
        return vmFactory(
            box.value,
            childrenFactory,
            hub,
            dispatcher,
            _name,
            _hint,
            _eagerChildren
        )
    }
}

// ── HierarchicalVM entry point ────────────────────────────────────────────────

extension HierarchicalVM {
    /// Entry point for the immutable fluent builder.
    ///
    /// ```swift
    /// let node = try HierarchicalVM<String, MyNode>.builder()
    ///     .model("root")
    ///     .childrenFactory { _ in [] }
    ///     .withDefaultServices()
    ///     .vmFactory { model, cf, hub, disp, name, hint, eager in
    ///         MyNode(model: model, childrenFactory: cf, hub: hub,
    ///                dispatcher: disp, name: name, hint: hint,
    ///                eagerChildren: eager)
    ///     }
    ///     .build()
    /// ```
    public static func builder() -> HierarchicalVMBuilder<TModel, TVM> {
        HierarchicalVMBuilder<TModel, TVM>()
    }
}
